Softeq.XToolkit.Common/Extensions/TaskExtensions.cs

```diff
@@ -11,43 +11,79 @@ namespace Softeq.XToolkit.Common.Extensions
+        /// </returns>
+        public static async Task WithTimeout(this Task task, TimeSpan timeout)
         {
-            var retTask = await Task.WhenAny(task, Task.Delay((int) timeout.TotalMilliseconds))
-                .ConfigureAwait(false);
+            // Details:
+            //     https://devblogs.microsoft.com/pfxteam/crafting-a-task-timeoutafter-method/
+            // Source:
+            //     https://github.com/microsoft/vs-threading/blob/master/src/Microsoft.VisualStudio.Threading/TplExtensions.cs#L73
 
-            if (retTask is Task<T>)
+            if (task == null)
             {
-                return task.Result;
+                throw new ArgumentNullException(nameof(task));
             }
 
-            return default;
+            using (var timerCancellation = new CancellationTokenSource())
+            {
+                var timeoutTask = Task.Delay(timeout, timerCancellation.Token);
+                var firstCompletedTask = await Task.WhenAny(task, timeoutTask).ConfigureAwait(false);
+
+                if (firstCompletedTask == timeoutTask)
+                {
+                    throw new TimeoutException();
+                }
+
+                timerCancellation.Cancel();
+
+                await task.ConfigureAwait(false);
+            }
         }
 
-        public static Task<T> SafeTaskWrapper<T>(this Task<T> task, ILogger logger = null)
+        /// <summary>
+        ///     Returns a task that completes as the original task completes or when a timeout expires, whichever happens first.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the original task.</typeparam>
+        /// <param name="task">The task to wait for.</param>
+        /// <param name="timeout">The maximum time to wait.</param>
+        /// <returns>
+        ///     A task that completes with the result of the specified <paramref name="task"/> or
+        ///     faults with a <see cref="TimeoutException"/> if <paramref name="timeout"/> elapses first.
+        /// </returns>
+        public static async Task<T> WithTimeout<T>(this Task<T> task, TimeSpan timeout)
         {
-            task.ContinueWith(t => { LogException(t, logger); }, CancellationToken.None,
-                TaskContinuationOptions.NotOnRanToCompletion, TaskScheduler.Default);
-            return task;
+            await WithTimeout((Task) task, timeout).ConfigureAwait(false);
+            return task.GetAwaiter().GetResult();
         }
 
-        public static Task SafeTaskWrapper(this Task task, ILogger logger = null)
+        /// <summary>
+        ///     Useful for fire-and-forget calls to async methods.
+        /// </summary>
+        /// <param name="task">The task.</param>
+        /// <param name="logger">Logger object.</param>
+        /// <returns>The task.</returns>
+        public static void FireAndForget(this Task task, ILogger logger = null)
```

Comment:
The name might be a bit confusing. The task is not actually fired here (it might have been fired a long ago). Here we just make sure that already fired task is handled properly. This name is better suited for Func<Task> actually (although we cant make an extension for it)

Reply:
Reimplemented.